### PR TITLE
14 move changes from crypro3 zk up to commit 5786b7bd. Changes include kzg commitment scheme, optimizations in lookup argument, zkllvm compatibility.

### PR DESCRIPTION
**Pull request by martun**
Move changes from crypro3 -zk up to commit 5786b7bd.

Changes include kzg commitment scheme, optimizations in lookup argument, zkllvm/EVM compatibility.

### CODE REVIEW

File: include/nil/actor/zk/commitments/polynomial/kzg.hpp [hunk 1]
```diff
@@ -30,101 +30,786 @@
...
                     using proof_type = commitment_type;
 
                     struct params_type {
-                        std::size_t a;
+                        using commitment_type = typename curve_type::template g1_type<>::value_type;
+                        using field_type = typename curve_type::scalar_field_type;
+
+                        single_commitment_type commitment_key;
+                        verification_key_type verification_key;
+                        params_type() {}
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/polynomial/kzg.hpp:91]: Why we need such constructor? It's default as I understand. So mb `= default;`

File: include/nil/actor/zk/commitments/polynomial/kzg.hpp [hunk 2]
```diff
@@ -30,101 +30,786 @@
...
+                            commitment_key(ck), verification_key(vk) {}
+                    };
+                    struct public_key_type {
+                        commitment_type commit;
+                        scalar_value_type z;
+                        scalar_value_type eval;
+                        public_key_type() {}
+                        public_key_type(commitment_type c, scalar_value_type z, scalar_value_type e)
+                                    : commit(c), z(z), eval(e) {}
+                        public_key_type operator=(const public_key_type &other) {
```
Comment (Zerg1996) [inline include/nil/actor/zk/commitments/polynomial/kzg.hpp:121]: Is it default behaviour for "="?

File: include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp [hunk 3]
```diff
@@ -32,34 +32,42 @@ namespace nil {
     namespace actor {
         namespace zk {
             namespace snark {
+                constexpr std::size_t FIXED_VALUES_BATCH = 0;
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp:35]: Looks like good work for enum :)

File: include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp [hunk 4]
```diff
@@ -32,34 +32,42 @@ namespace nil {
...
-                 * A proof for the Placeholder cheme.
+                 * A proof for the Placeholder scheme.
                  *
                  * While the proof has a structure, externally one merely opaquely produces,
                  * serializes/deserializes, and verifies proofs. We only expose some information
                  * about the structure for marshalling purposes.
                  */
                 template<typename FieldType, typename ParamsType>
                 struct placeholder_proof {
+                    static constexpr std::size_t FIXED_VALUES_BATCH = 0;
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp:50]: I think here also enum preferable. Also if we define them in `actor::zk::snark` namespace - we can reuse them here.

File: include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp [hunk 5]
```diff
@@ -339,26 +317,25 @@ namespace nil {
                         F[4] = lookup_argument[1];
                         F[5] = lookup_argument[2];
                         F[6] = lookup_argument[3];
-                        F[7] = lookup_argument[4];
```
Comment (Zerg1996) [inline include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp:342]: Can we simple define `F` as F = {permutation_argument[0], permutation_argument[1], ...}

File: test/commitment/fri.cpp [hunk 6]
```diff
@@ -95,7 +95,7 @@ ACTOR_THREAD_TEST_CASE(fri_basic_test) {
     constexpr static const std::size_t lambda = 40;
     constexpr static const std::size_t batches_num = 1;
 
-    typedef zk::commitments::fri<FieldType, merkle_hash_type, transcript_hash_type, lambda, m, batches_num> fri_type;
+    typedef zk::commitments::fri<FieldType, merkle_hash_type, transcript_hash_type, lambda, m, true> fri_type;
```
Comment (Zerg1996) [inline test/commitment/fri.cpp:98]: Does that mean we have batch size equal 1?

File: test/systems/plonk/placeholder/performance.cpp [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+    placeholder_fibonacci_params::m
+>;
+
+using lpc_type = commitments::list_polynomial_commitment<field_type, lpc_params_type>;
+using lpc_scheme_type = typename commitments::lpc_commitment_scheme<lpc_type>;
+using lpc_placeholder_params_type = nil::crypto3::zk::snark::placeholder_params<circuit_fib_params, lpc_scheme_type>;
+using policy_type = zk::snark::detail::placeholder_policy<field_type, lpc_placeholder_params_type>;
+
+/*
+ACTOR_THREAD_TEST_CASE(small_merkle_tree_test1) {
```
Comment (Zerg1996) [inline test/systems/plonk/placeholder/performance.cpp:174]: Let's remove it?

**Review by Zerg1996 — APPROVED** [review]
I didn't see any `.get()` calls. For actor side this PR simple renaming and merge diff. So I can approve it.

File: include/nil/actor/zk/commitments/polynomial/fri.hpp [hunk 8]
```diff
@@ -94,47 +99,54 @@ namespace nil {
...
+                static typename FRI::basic_fri::proof_type proof_eval(
+                    PolynomialType &g,
+                    typename FRI::basic_fri::merkle_tree_type &tree,
+                    const typename FRI::params_type &fri_params,
+                    typename FRI::transcript_type &transcript = typename FRI::transcript_type()
+                ){  
+                    std::map<std::size_t, std::vector<PolynomialType>> gs;
+                    gs[0]={g};
+                    std::map<std::size_t, typename FRI::basic_fri::merkle_tree_type> trees;
+                    trees[0] = typename FRI::basic_fri::merkle_tree_type(tree);;
```
[the remark this answers is not in the document]
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/fri.hpp:123]: Fixed.

File: include/nil/actor/zk/commitments/polynomial/kzg.hpp (hunk 2 above)
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/kzg.hpp:121]: Fixed.

File: include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp (hunk 3 above)
Comment (martun) [reply include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp:35]: Let's change this later, these constants are copied twice in this code, and used all over the place, no time to change this now.

File: include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp (hunk 4 above)
Comment (martun) [reply include/nil/actor/zk/snark/systems/plonk/placeholder/proof.hpp:50]: These are the same values, we are supposed to remove the upper one. Will do this sometime later.

File: include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp (hunk 5 above)
Comment (martun) [reply include/nil/actor/zk/snark/systems/plonk/placeholder/verifier.hpp:342]: Yes, fixed.

File: test/commitment/fri.cpp (hunk 6 above)
Comment (martun) [reply test/commitment/fri.cpp:98]: Nope, Elena changes the commitments totally. Now this "true" means "use grinding", and there are not batches number any more as far as I understand.

File: test/systems/plonk/placeholder/performance.cpp (hunk 7 above)
Comment (martun) [reply test/systems/plonk/placeholder/performance.cpp:174]: I have refactored this code in the crypto3-zk. I will move them tomorrow, and we will have a good code here.

File: include/nil/actor/zk/commitments/polynomial/kzg.hpp (hunk 1 above)
Comment (martun) [reply include/nil/actor/zk/commitments/polynomial/kzg.hpp:91]: Fixed.

**Pull request merged**